### PR TITLE
Add service doc generation for grpc services.

### PR DESCRIPTION
**Pull request by anuraaga**
Adds a DocPlugin that extracts service information from grpc protobuf descriptors. Can optionally find docstrings from descriptor files compiled with source code info.

Also:
- Extracts common reflection logic for docstring extractors
- Forks testing protos from grpc-testing to allow compiling with source code info and for future extension
- Fixes double-escaping of docstrings by using SafeString (and also using Handlebars built-in escape method).

### CODE REVIEW

File: grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcDocServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                .build();
+        sb.serviceAt("/", simpleService);
+        sb.serviceUnder("/docs/", new DocService().decorate(LoggingService::new));
+    }
+
+    @Test
+    public void testOk() throws Exception {
+        System.out.println(httpPort());
+        while (true) {
+            Thread.sleep(Long.MAX_VALUE);
```
Comment (trustin) [inline grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcDocServiceTest.java]: I guess this will make Travis unhappy ;-)

Comment (anuraaga) [inline grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcDocServiceTest.java]: Luckily I have checkstyle errors :D

**Comment by anuraaga** [comment]
@trustin This PR is basically ready for review - tests and common logic extraction are done. I've left it WIP though since there will probably be reasonably big changes after #422 is merged.

**Comment by anuraaga** [comment]
@trustin This is ready for review :D

File: core/src/main/java/com/linecorp/armeria/server/docs/DocStringExtractor.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * A supporting base class for implementing the standard pattern of extracting docstrings
+ * from arbitrary files in a particular classpath location.
+ */
+public abstract class DocStringExtractor {
+
+    private static final Logger logger = LoggerFactory.getLogger(DocStringExtractor.class);
+
+    private static final Map<ClassLoader, Map<String, String>> cached = new ConcurrentHashMap<>();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocStringExtractor.java]: I wonder if we should use weak keys. What do you think?

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcDocStringExtractor.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+ *
+ * <p>To include docstrings in {@link com.linecorp.armeria.server.docs.DocService} pages,
+ * configure the protobuf compiler to generate a descriptor set with source info and all imports
+ * included. Place the descriptor set in the classpath location {@code META-INF/armeria/grpc} and ensure
+ * the file extension is '.dsc'. The classpath location can be changed by setting the
+ * {@code com.linecorp.armeria.grpc.descriptorDir} system property.
+ *
+ * <p>For example, to generate a descriptor set in gradle:
+ * <pre>
+ *     {@code
```
Comment (trustin) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcDocStringExtractor.java]: Could be merged into a single line. i.e. `<pre>{@code`

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcDocStringExtractor.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+ *                 task.generateDescriptorSet = true
+ *                 task.descriptorSetOptions.includeSourceInfo = true
+ *                 task.descriptorSetOptions.includeImports = true
+ *                 task.descriptorSetOptions.path =
+ *                         "${buildDir}/resources/main/META-INF/armeria/grpc/service-name.dsc"
+ *             }
+ *         }
+ *     }
+ *     }
+ * </pre>
```
Comment (trustin) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcDocStringExtractor.java]: Could be emrged into a single line. i.e. `}</pre>`

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcDocStringExtractor.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+ * included. Place the descriptor set in the classpath location {@code META-INF/armeria/grpc} and ensure
+ * the file extension is '.dsc'. The classpath location can be changed by setting the
+ * {@code com.linecorp.armeria.grpc.descriptorDir} system property.
+ *
+ * <p>For example, to generate a descriptor set in gradle:
+ * <pre>
+ *     {@code
+ *     protobuf {
+ *         generateProtoTasks {
+ *             all().each { task ->
```
Comment (trustin) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcDocStringExtractor.java]: This example block could be indented one less level.

File: core/src/main/java/com/linecorp/armeria/server/docs/DocStringExtractor.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+        return getDocStringsFromFiles(files);
+    }
+
+    /**
+     * Determine whether the file at {@code filename} should be processed by the {@link DocStringExtractor}.
+     * This will usually look at the file extension, but the default implementation can be used to check all
+     * files at a particular path.
+     */
+    protected boolean processFile(String filename) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocStringExtractor.java]: `acceptFile`?

File: core/src/main/java/com/linecorp/armeria/server/docs/DocStringExtractor.java (hunk 2 above)
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocStringExtractor.java]: Sounds like a good idea, how about I come back to this after we add Caffeine, which will make it trivial?

Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocStringExtractor.java]: SGTM

**Comment by trustin** [comment]
LGTM

**Pull request merged**